lib/runtime-core/src/typed_func.rs

```diff
@@ -438,33 +472,91 @@ macro_rules! impl_traits {
             }
         }
 
-        impl< $( $x, )* Rets, Trap, FN > ExternalFunction<( $( $x ),* ), Rets> for FN
+        impl< $( $x, )* Rets, Trap, FN > ExternalFunction<ExplicitVmCtx, ( $( $x ),* ), Rets> for FN
+        where
+            $( $x: WasmExternType, )*
+            Rets: WasmTypeList,
+            Trap: TrapEarly<Rets>,
+            FN: Fn(&mut vm::Ctx $( , $x )*) -> Trap,
+        {
+            #[allow(non_snake_case)]
+            fn to_raw(&self) -> NonNull<vm::Func> {
+                if mem::size_of::<Self>() == 0 {
+                    /// This is required for the llvm backend to be able to unwind through this function.
+                    #[cfg_attr(nightly, unwind(allowed))]
+                    extern fn wrap<$( $x, )* Rets, Trap, FN>(
+                        vmctx: &mut vm::Ctx $( , $x: <$x as WasmExternType>::Native )*
+                    ) -> Rets::CStruct
+                    where
+                        $( $x: WasmExternType, )*
+                        Rets: WasmTypeList,
+                        Trap: TrapEarly<Rets>,
+                        FN: Fn(&mut vm::Ctx, $( $x, )*) -> Trap,
+                    {
+                        let f: FN = unsafe { mem::transmute_copy(&()) };
+
+                        let err = match panic::catch_unwind(
+                            panic::AssertUnwindSafe(
+                                || {
+                                    f(vmctx $( , WasmExternType::from_native($x) )* ).report()
+                                }
+                            )
+                        ) {
+                            Ok(Ok(returns)) => return returns.into_c_struct(),
+                            Ok(Err(err)) => {
+                                let b: Box<_> = err.into();
+                                b as Box<dyn Any>
+                            },
+                            Err(err) => err,
+                        };
+
+                        unsafe {
+                            (&*vmctx.module).runnable_module.do_early_trap(err)
+                        }
+                    }
+
+                    NonNull::new(wrap::<$( $x, )* Rets, Trap, Self> as *mut vm::Func).unwrap()
+                } else {
+                    assert_eq!(
+                        mem::size_of::<Self>(),
+                        mem::size_of::<usize>(),
+                        "you cannot use a closure that captures state for `Func`."
+                    );
+
+                    NonNull::new(unsafe {
+                        mem::transmute_copy::<_, *mut vm::Func>(self)
+                    }).unwrap()
+                }
+            }
+        }
+
+        impl< $( $x, )* Rets, Trap, FN > ExternalFunction<ImplicitVmCtx, ( $( $x ),* ), Rets> for FN
```

Comment:
Implementation of `ExternalFunction<ImplicitVmCtx, …>`, where `FN` must be `Fn(…)` (without `vm::Ctx`). The `wrap` function _still_ has a `&mut vm::Ctx` argument! Why? Because the backend always inserts a pointer to `vm::Ctx` in the stack, so `wrap` can read it.

Why grabbing `vm::Ctx` then? It is mandatory to trap error with `do_early_trap`.